File: backend/app/models/practice.py
from pydantic import BaseModel, Field
from typing import Optional, List, Any, Dict
from datetime import datetime
from enum import Enum
from uuid import UUID

from .problem import ProblemType


class SubmissionResult(str, Enum):
    """Submission result types."""
    CORRECT = "correct"
    INCORRECT = "incorrect"
    PARTIAL = "partial"
    ERROR = "error"


class TestResult(BaseModel):
    """Result of a single test case."""
    test_id: str
    passed: bool
    input: str
    expected: str
    actual: Optional[str] = None
    error: Optional[str] = None


class BlankSubmission(BaseModel):
    """Submission for blank-fill problems."""
    problem_id: UUID
    answers: Dict[str, str]  # blank_id -> answer


class PuzzleBlockSubmission(BaseModel):
    """Single block in puzzle submission."""
    id: str
    indentation: int


class PuzzleSubmission(BaseModel):
    """Submission for puzzle (Parsons) problems."""
    problem_id: UUID
    block_order: List[PuzzleBlockSubmission]  # Blocks in user's arranged order with indentation


class NewBadge(BaseModel):
    """Newly earned badge info."""
    code: str
    name: str
    icon_url: Optional[str] = None
    rarity: str = "common"


class SubmissionResponse(BaseModel):
    """Response for problem submission."""
    result: SubmissionResult
    is_correct: bool
    score: Optional[int] = None
    xp_earned: int = 0
    feedback: Optional[str] = None

    # Type-specific results
    blank_results: Optional[Dict[str, bool]] = None  # blank_id -> correct
    puzzle_results: Optional[Dict[str, bool]] = None  # block_id -> correct position/indentation

    # Newly earned badges
    new_badges: Optional[List[NewBadge]] = None


class CodeExecutionRequest(BaseModel):
    """Request for code execution."""
    code: str
    language: str = "javascript"
    test_input: Optional[str] = None


class CodeExecutionResponse(BaseModel):
    """Response from code execution."""
    success: bool
    output: Optional[str] = None
    error: Optional[str] = None
    execution_time: Optional[int] = None  # ms


class PracticeSession(BaseModel):
    """Practice session model."""
    id: UUID
    user_id: UUID
    problem_id: UUID
    problem_type: ProblemType
    started_at: datetime
    completed_at: Optional[datetime] = None
    is_correct: Optional[bool] = None
    score: Optional[int] = None
    hints_used: int = 0
    xp_earned: int = 0


class AttemptCreate(BaseModel):
    """Model for creating an attempt record."""
    user_id: UUID
    problem_id: UUID
    is_correct: bool
    score: Optional[int] = None
    submitted_code: Optional[str] = None
    submitted_answer: Optional[str] = None
    time_spent: Optional[int] = None
    hints_used: int = 0
    xp_earned: int = 0


class Attempt(BaseModel):
    """Attempt record model."""
    id: UUID
    user_id: UUID
    problem_id: UUID
    is_correct: bool
    score: Optional[int] = None
    submitted_code: Optional[str] = None
    submitted_answer: Optional[str] = None
    started_at: Optional[datetime] = None
    submitted_at: datetime
    time_spent: Optional[int] = None
    hints_used: int = 0
    xp_earned: int = 0


class XPConfig:
    """XP configuration for different problem types and difficulties."""

    # 난이도별 XP 획득량
    DIFFICULTY_XP = {
        "easy": 40,
        "medium": 60,
        "medium_hard": 90,
        "hard": 120,
        "very_hard": 200,
    }

    # 문제 유형별 기본 XP (난이도 없을 때 fallback)
    BLANK_CORRECT = 40     # 빈칸 채우기 기본
    PUZZLE_CORRECT = 40    # 퍼즐(Parsons) 기본
    GUIDED_CORRECT = 60    # 가이드 문제 기본

    # 힌트 페널티
    HINT_COST = 5          # 힌트 사용 시 -5 XP

    @classmethod
    def get_xp_for_difficulty(cls, difficulty: str, problem_type: str = "blank") -> int:
        """난이도에 따른 XP 반환, 없으면 문제 유형별 기본값"""
        if difficulty in cls.DIFFICULTY_XP:
            return cls.DIFFICULTY_XP[difficulty]
        # fallback to problem type default
        type_defaults = {
            "blank": cls.BLANK_CORRECT,
            "puzzle": cls.PUZZLE_CORRECT,
            "guided": cls.GUIDED_CORRECT,
        }
        return type_defaults.get(problem_type, 40)


class RecordSubmission(BaseModel):
    """Simple submission for recording problem solve (XP & grass)."""
    problem_id: str  # UUID or generated problem ID
    base_problem_id: Optional[str] = None  # base_problems 테이블의 UUID (잔디 클릭 시 문제 정보 표시용)
    problem_type: str = "blank"  # blank, puzzle, guided
    difficulty: Optional[str] = None  # easy, medium, medium_hard, hard, very_hard
    is_correct: bool = True
    xp_earned: Optional[int] = None  # None이면 난이도 기반 자동 계산
    problem_name: Optional[str] = None  # 문제 이름 (잔디 클릭 시 표시용)
    hints_used: Optional[int] = None  # 힌트 사용 횟수 (기록용)
    topics: Optional[List[str]] = None  # 문제 주제/태그 (예: ["DP", "그래프"])
    attempt_id: Optional[str] = None  # 기존 pending attempt ID (attempt_details 기록용)
    # 추가 필드: attempts 테이블 완전 저장용
    time_spent: Optional[int] = None  # 풀이 시간 (초)
    submitted_code: Optional[str] = None  # 제출한 코드 (guided/코드 제출용)
    session_id: Optional[str] = None  # 세션 ID (세션 추적용)
    chat_session_id: Optional[str] = None  # 🚀 chat_sessions 테이블 ID (대화 기록 조회용)

    # ============================================================
    # 상세 기록용 필드 (attempt_details 테이블)
    # ============================================================
    # Blank 유형: 사용자 답변 정보
    blank_answers: Optional[Dict[str, str]] = None  # {blank_id: user_answer}
    blank_correct_answers: Optional[Dict[str, str]] = None  # {blank_id: correct_answer}
    blank_results: Optional[Dict[str, bool]] = None  # {blank_id: is_correct}

    # Puzzle 유형: 블록 순서 정보
    puzzle_user_order: Optional[List[str]] = None  # 사용자가 배치한 블록 순서
    puzzle_correct_order: Optional[List[str]] = None  # 정답 블록 순서
    puzzle_wrong_positions: Optional[List[int]] = None  # 틀린 위치들


class FeedbackData(BaseModel):
    """피드백 데이터 (record_solve 응답용)."""
    grade: str = "learning"  # perfect, excellent, good, keep_going, learning
    grade_emoji: str = "🌱"
    grade_message: str = ""
    summary_title: str = ""
    summary_highlight: str = ""
    efficiency_score: int = 0
    speed_score: int = 0
    understanding_score: int = 0
    learning_points: List[str] = []
    improvements: List[str] = []
    encouragement: str = ""


class SeedAwarded(BaseModel):
    """씨앗 보상 정보."""
    seed_code: str      # "seed_carrot"
    crop_code: str      # "carrot"
    rarity: str         # "common", "uncommon", "rare"
    crop_name_ko: str   # "당근"


class RecordResponse(BaseModel):
    """Response for record submission."""
    success: bool
    xp_earned: int
    message: str
    new_badges: Optional[List[NewBadge]] = None
    feedback: Optional[FeedbackData] = None  # 피드백 데이터 (정답 시)
    solve_count: Optional[int] = None  # 문제 풀이 수 (첫 정답 시에만 증가)
    seed_awarded: Optional[SeedAwarded] = None  # 씨앗 보상 (첫 정답 시에만)


class HintCheckResponse(BaseModel):
    """Response for hint availability check."""
    can_use: bool
    current_xp: int
    hint_cost: int
    message: str


class HintUseResponse(BaseModel):
    """Response after using hint."""
    success: bool
    xp_deducted: int
    remaining_xp: int
    message: str


# ============================================================
# Start Practice (Attempt Tracking)
# ============================================================

class StartPracticeRequest(BaseModel):
    """문제 풀이 시작 요청 - attempt 생성"""
    problem_id: str  # UUID or generated problem ID
    problem_type: str = "blank"  # blank, puzzle, guided
    difficulty: Optional[str] = None
    problem_name: Optional[str] = None
    topics: Optional[List[str]] = None
    chat_session_id: Optional[str] = None  # chat_sessions 테이블 연동용


class StartPracticeResponse(BaseModel):
    """문제 풀이 시작 응답"""
    attempt_id: str
    started_at: str
    message: str
    session_id: Optional[str] = None  # 세션 추적용 ID


# ============================================================
# Session Tracking Models
# ============================================================

class SessionHeartbeatRequest(BaseModel):
    """세션 하트비트 요청"""
    session_id: str
    hints_used: Optional[int] = None
    attempt_count: Optional[int] = None


class SessionHeartbeatResponse(BaseModel):
    """세션 하트비트 응답"""
    success: bool
    session_status: str  # active, expired, abandoned
    time_spent: Optional[int] = None
    error: Optional[str] = None


class SessionEndRequest(BaseModel):
    """세션 종료 요청 (완료/스킵/포기)"""
    session_id: str
    end_type: str  # complete, skip, abandon
    reason: Optional[str] = None
    is_correct: Optional[bool] = None
    score: Optional[int] = None
    chat_session_id: Optional[str] = None  # 🚀 대화 기록 조회용


class SessionEndResponse(BaseModel):
    """세션 종료 응답"""
    success: bool
    session_id: str
    status: str  # completed, skipped, abandoned
    time_spent: int
    hints_used: int = 0
    attempt_count: int = 0
    error: Optional[str] = None


# ============================================================
# Attempt Detail Recording
# ============================================================

class AttemptDetailAction(str, Enum):
    """attempt_details action types"""
    BLANK_SUBMIT = "blank_submit"
    BLANK_HINT_REQUEST = "blank_hint_request"
    BLANK_HINT_REVEAL = "blank_hint_reveal"
    PUZZLE_SUBMIT = "puzzle_submit"
    PUZZLE_HINT_REQUEST = "puzzle_hint_request"
    GUIDED_MESSAGE = "guided_message"
    GUIDED_STEP_COMPLETE = "guided_step_complete"
    GIVE_UP = "give_up"


class RecordAttemptDetailRequest(BaseModel):
    """attempt_details 기록 요청 (내부용)"""
    attempt_id: str
    action_type: str
    step_number: Optional[int] = None

    # Blank specific
    blank_index: Optional[int] = None
    blank_user_answer: Optional[str] = None
    blank_correct_answer: Optional[str] = None
    blank_is_correct: Optional[bool] = None
    blank_hint_level: Optional[int] = None
    blank_hint_content: Optional[str] = None

    # Puzzle specific
    puzzle_user_order: Optional[List[int]] = None
    puzzle_correct_order: Optional[List[int]] = None
    puzzle_wrong_positions: Optional[List[Dict]] = None
    puzzle_hint_content: Optional[str] = None

    # Guided specific
    guided_step: Optional[int] = None
    guided_user_message: Optional[str] = None
    guided_tutor_response: Optional[str] = None
    guided_understanding_score: Optional[float] = None

    # Hint tracking
    hint_was_requested: bool = False

File: src/lib/api/practice.ts
/**
 * Practice API Functions
 */

import { api } from './client';

export interface BlankSubmission {
  problemId: string;
  answers: Record<string, string>;
}

export interface PuzzleSubmission {
  problemId: string;
  blockOrder: Array<{ id: string; indentation: number }>;
}

export interface NewBadge {
  code: string;
  name: string;
  iconUrl?: string | null;
  rarity: string;
}

export interface SeedAwarded {
  seedCode: string;      // "seed_carrot"
  cropCode: string;      // "carrot"
  rarity: string;        // "common", "uncommon", "rare"
  cropNameKo: string;    // "당근"
}

export interface BlankResult {
  results: Record<string, boolean>;
  allCorrect: boolean;
  xpEarned: number;
  newBadges?: NewBadge[];
}

export interface PuzzleResult {
  isCorrect: boolean;
  results: Record<string, boolean>;
  xpEarned: number;
  newBadges?: NewBadge[];
}

export interface RunCodeResult {
  output?: string;
  error?: string;
}

export interface RecordSubmission {
  problemId: string;
  baseProblemId?: string;  // base_problems 테이블 UUID (잔디 클릭 시 문제 정보 표시용)
  problemType: 'blank' | 'puzzle' | 'guided';
  isCorrect: boolean;
  xpEarned?: number;
  problemName?: string;  // 문제 이름 (잔디 클릭 시 표시용)
  hintsUsed?: number;    // 힌트 사용 횟수 (기록용)
  topics?: string[];     // 문제 주제/태그 (예: ["DP", "그래프"])
  attemptId?: string;    // 기존 pending attempt ID (attempt_details 기록용)
  // ✅ attempts 테이블 완전 저장용 추가 필드
  timeSpent?: number;    // 풀이 소요 시간 (초)
  submittedCode?: string; // 제출한 코드 (guided/코드 제출용)
  sessionId?: string;    // 세션 ID (세션 추적용)

  // ============================================================
  // 상세 기록용 필드 (attempt_details 테이블)
  // ============================================================
  // Blank 유형: 사용자 답변 정보
  blankAnswers?: Record<string, string>;      // {blank_id: user_answer}
  blankCorrectAnswers?: Record<string, string>; // {blank_id: correct_answer}
  blankResults?: Record<string, boolean>;     // {blank_id: is_correct}

  // Puzzle 유형: 블록 순서 정보
  puzzleUserOrder?: string[];     // 사용자가 배치한 블록 순서
  puzzleCorrectOrder?: string[];  // 정답 블록 순서
  puzzleWrongPositions?: number[]; // 틀린 위치들
}

export interface RecordResult {
  success: boolean;
  xpEarned: number;
  message: string;
  newBadges?: NewBadge[];
  solveCount?: number;  // 문제 풀이 수 (첫 정답 시에만 증가)
  seedAwarded?: SeedAwarded;  // 씨앗 보상 (첫 정답 시에만)
}

export interface HintCheckResult {
  canUse: boolean;
  currentXp: number;
  hintCost: number;
  message: string;
}

export interface HintUseResult {
  success: boolean;
  xpDeducted: number;
  remainingXp: number;
  message: string;
}

// ============================================================
// 힌트 요청/응답 타입 (Blank, Puzzle, Guided 공통 구조)
// ============================================================

export interface BlankHintRequest {
  problemId: string;
  blankIndex: number;
  codeTemplate?: string;
  attemptId?: string;  // attempt 추적용
}

export interface BlankHintResult {
  blankIndex: number;
  hintContent: string;
  fromCache: boolean;
}

export interface PuzzleHintRequest {
  problemId: string;
  blockIndex: number;
  blocks?: Array<{ id: string; code: string }>;
  attemptId?: string;  // attempt 추적용
}

export interface PuzzleHintResult {
  blockIndex: number;
  hintContent: string;
  fromCache: boolean;
}

export interface GuidedHintRequest {
  problemId: string;
  stepIndex: number;
  steps?: Array<unknown>;
  attemptId?: string;  // attempt 추적용
}

export interface GuidedHintResult {
  stepIndex: number;
  hintContent: string;
  fromCache: boolean;
}

// ============================================================
// Guided Tutor Chat (1대1 대화형 튜터)
// ============================================================

export interface GuidedTutorStartRequest {
  problemId: string;
  sessionId?: string;
}

export interface GuidedTutorChatRequest {
  problemId: string;
  message: string;
  userCode?: string;
  sessionId?: string;
  conversationHistory?: Array<{ role: 'user' | 'assistant'; content: string }>;
}

export interface InitialGuideData {
  introduction: string;
  conceptsExplanation: Array<{
    concept: string;
    explanation: string;
    whyNeeded: string;
  }>;
  variablesGuide: {
    totalCount: number;
    variables: Array<{
      role: string;
      type: string;
      initialValue: string;
      suggestedName: string;
      explanation: string;
    }>;
  };
  flowOverview: string[];
  firstStep: {
    stepName: string;
    instruction: string;
    starterQuestion: string;
  };
}

export interface GuidedTutorChatResponse {
  response: string;
  isInitialGuide: boolean;
  isComplete: boolean;
  studentProgress?: {
    understandingScore: number;
    conceptsMastered: string[];
    conceptsStruggling: string[];
    currentFocus: string;
    hintsReceived: number;
    attempts: number;
  };
  initialGuide?: InitialGuideData;
  suggestedNextStep?: string;
}

// ============================================================
// Start Practice (Attempt Tracking)
// ============================================================

export interface StartPracticeRequest {
  problemId: string;
  problemType: 'blank' | 'puzzle' | 'guided';
  difficulty?: string;
  problemName?: string;
  topics?: string[];
}

export interface StartPracticeResult {
  attemptId: string;
  startedAt: string;
  message: string;
  sessionId?: string;  // 세션 추적용 ID
}

// ============================================================
// Session Tracking
// ============================================================

export interface SessionHeartbeatRequest {
  sessionId: string;
  hintsUsed?: number;
  attemptCount?: number;
}

export interface SessionHeartbeatResult {
  success: boolean;
  sessionStatus: 'active' | 'expired' | 'abandoned';
  timeSpent?: number;
  error?: string;
}

export interface SessionEndRequest {
  sessionId: string;
  endType: 'complete' | 'skip' | 'abandon';
  reason?: string;
  isCorrect?: boolean;
  score?: number;
}

export interface SessionEndResult {
  success: boolean;
  sessionId: string;
  status: 'completed' | 'skipped' | 'abandoned';
  timeSpent: number;
  hintsUsed: number;
  attemptCount: number;
  error?: string;
}

export const practiceApi = {
  /**
   * Start practice session - creates pending attempt
   */
  async startPractice(request: StartPracticeRequest): Promise<StartPracticeResult> {
    const response = await api.post<{
      attempt_id: string;
      started_at: string;
      message: string;
      session_id?: string;
    }>('/practice/start', {
      problem_id: request.problemId,
      problem_type: request.problemType,
      difficulty: request.difficulty,
      problem_name: request.problemName,
      topics: request.topics,
    });
    if (response.error) throw new Error(response.error.message);
    return {
      attemptId: response.data!.attempt_id,
      startedAt: response.data!.started_at,
      message: response.data!.message,
      sessionId: response.data!.session_id,
    };
  },

  /**
   * Send session heartbeat to keep session alive
   */
  async sessionHeartbeat(request: SessionHeartbeatRequest): Promise<SessionHeartbeatResult> {
    const response = await api.post<{
      success: boolean;
      session_status: 'active' | 'expired' | 'abandoned';
      time_spent?: number;
      error?: string;
    }>('/practice/session/heartbeat', {
      session_id: request.sessionId,
      hints_used: request.hintsUsed,
      attempt_count: request.attemptCount,
    });
    if (response.error) throw new Error(response.error.message);
    return {
      success: response.data!.success,
      sessionStatus: response.data!.session_status,
      timeSpent: response.data!.time_spent,
      error: response.data!.error,
    };
  },

  /**
   * End session (complete, skip, or abandon)
   */
  async sessionEnd(request: SessionEndRequest): Promise<SessionEndResult> {
    const response = await api.post<{
      success: boolean;
      session_id: string;
      status: 'completed' | 'skipped' | 'abandoned';
      time_spent: number;
      hints_used: number;
      attempt_count: number;
      error?: string;
    }>('/practice/session/end', {
      session_id: request.sessionId,
      end_type: request.endType,
      reason: request.reason,
      is_correct: request.isCorrect,
      score: request.score,
    });
    if (response.error) throw new Error(response.error.message);
    return {
      success: response.data!.success,
      sessionId: response.data!.session_id,
      status: response.data!.status,
      timeSpent: response.data!.time_spent,
      hintsUsed: response.data!.hints_used,
      attemptCount: response.data!.attempt_count,
      error: response.data!.error,
    };
  },

  /**
   * Run code without submitting
   */
  async runCode(code: string, language: string = 'python'): Promise<RunCodeResult> {
    const response = await api.post<RunCodeResult>('/practice/run', {
      code,
      language,
    });
    if (response.error) throw new Error(response.error.message);
    return response.data!;
  },

  /**
   * Submit blank fill answers
   */
  async submitBlank(submission: BlankSubmission): Promise<BlankResult> {
    const response = await api.post<{
      results: Record<string, boolean>;
      all_correct: boolean;
      xp_earned: number;
      new_badges?: Array<{ code: string; name: string; icon_url?: string; rarity: string }>;
    }>('/practice/submit/blank', {
      problem_id: submission.problemId,
      answers: submission.answers,
    });
    if (response.error) throw new Error(response.error.message);
    const data = response.data!;
    return {
      results: data.results,
      allCorrect: data.all_correct,
      xpEarned: data.xp_earned,
      newBadges: data.new_badges?.map(b => ({
        code: b.code,
        name: b.name,
        iconUrl: b.icon_url,
        rarity: b.rarity,
      })),
    };
  },

  /**
   * Submit puzzle (Parsons Problem) answer
   */
  async submitPuzzle(submission: PuzzleSubmission): Promise<PuzzleResult> {
    const response = await api.post<{
      is_correct: boolean;
      results: Record<string, boolean>;
      xp_earned: number;
      new_badges?: Array<{ code: string; name: string; icon_url?: string; rarity: string }>;
    }>('/practice/submit/puzzle', {
      problem_id: submission.problemId,
      block_order: submission.blockOrder,
    });
    if (response.error) throw new Error(response.error.message);
    const data = response.data!;
    return {
      isCorrect: data.is_correct,
      results: data.results,
      xpEarned: data.xp_earned,
      newBadges: data.new_badges?.map(b => ({
        code: b.code,
        name: b.name,
        iconUrl: b.icon_url,
        rarity: b.rarity,
      })),
    };
  },

  /**
   * Record problem solve for XP and grass (simple record without validation)
   */
  async recordSolve(submission: RecordSubmission & { difficulty?: string }): Promise<RecordResult> {
    const response = await api.post<{
      success: boolean;
      xp_earned: number;
      message: string;
      new_badges?: Array<{ code: string; name: string; icon_url?: string; rarity: string }>;
      solve_count?: number;  // 문제 풀이 수
      seed_awarded?: { seed_code: string; crop_code: string; rarity: string; crop_name_ko: string };  // 씨앗 보상
    }>('/practice/submit/record', {
      problem_id: submission.problemId,
      base_problem_id: submission.baseProblemId,
      problem_type: submission.problemType,
      difficulty: submission.difficulty,
      is_correct: submission.isCorrect,
      xp_earned: submission.xpEarned,
      problem_name: submission.problemName,
      hints_used: submission.hintsUsed,  // 힌트 사용 횟수
      topics: submission.topics,  // 문제 주제/태그
      attempt_id: submission.attemptId,  // 기존 pending attempt ID
      // ✅ attempts 테이블 완전 저장용 추가 필드
      time_spent: submission.timeSpent,  // 풀이 소요 시간 (초)
      submitted_code: submission.submittedCode,  // 제출한 코드
      session_id: submission.sessionId,  // 세션 ID
      // ✅ attempt_details 상세 기록용 필드
      blank_answers: submission.blankAnswers,
      blank_correct_answers: submission.blankCorrectAnswers,
      blank_results: submission.blankResults,
      puzzle_user_order: submission.puzzleUserOrder,
      puzzle_correct_order: submission.puzzleCorrectOrder,
      puzzle_wrong_positions: submission.puzzleWrongPositions,
    });
    if (response.error) throw new Error(response.error.message);
    const data = response.data!;
    return {
      success: data.success,
      xpEarned: data.xp_earned,
      message: data.message,
      newBadges: data.new_badges?.map(b => ({
        code: b.code,
        name: b.name,
        iconUrl: b.icon_url,
        rarity: b.rarity,
      })),
      solveCount: data.solve_count,  // 문제 풀이 수
      seedAwarded: data.seed_awarded ? {
        seedCode: data.seed_awarded.seed_code,
        cropCode: data.seed_awarded.crop_code,
        rarity: data.seed_awarded.rarity,
        cropNameKo: data.seed_awarded.crop_name_ko,
      } : undefined,
    };
  },

  /**
   * Check if user can use a hint (has enough XP)
   */
  async checkHint(): Promise<HintCheckResult> {
    const response = await api.get<{
      can_use: boolean;
      current_xp: number;
      hint_cost: number;
      message: string;
    }>('/practice/hint/check');
    if (response.error) throw new Error(response.error.message);
    return {
      canUse: response.data!.can_use,
      currentXp: response.data!.current_xp,
      hintCost: response.data!.hint_cost,
      message: response.data!.message,
    };
  },

  /**
   * Use a hint (deducts XP)
   */
  async useHint(): Promise<HintUseResult> {
    const response = await api.post<{
      success: boolean;
      xp_deducted: number;
      remaining_xp: number;
      message: string;
    }>('/practice/hint/use');
    if (response.error) throw new Error(response.error.message);
    return {
      success: response.data!.success,
      xpDeducted: response.data!.xp_deducted,
      remainingXp: response.data!.remaining_xp,
      message: response.data!.message,
    };
  },

  /**
   * Get blank hint (역할/이유 설명만, 정답 없음)
   */
  async getBlankHint(request: BlankHintRequest): Promise<BlankHintResult> {
    const response = await api.post<{
      blank_index: number;
      hint_content: string;
      from_cache: boolean;
    }>('/practice/hint/blank', {
      problem_id: request.problemId,
      blank_index: request.blankIndex,
      code_template: request.codeTemplate,
      attempt_id: request.attemptId,  // attempt 추적용
    });
    if (response.error) throw new Error(response.error.message);
    return {
      blankIndex: response.data!.blank_index,
      hintContent: response.data!.hint_content,
      fromCache: response.data!.from_cache,
    };
  },

  /**
   * Get puzzle hint (블록 역할 설명, 순서 알려주지 않음)
   */
  async getPuzzleHint(request: PuzzleHintRequest): Promise<PuzzleHintResult> {
    const response = await api.post<{
      block_index: number;
      hint_content: string;
      from_cache: boolean;
    }>('/practice/hint/puzzle', {
      problem_id: request.problemId,
      block_index: request.blockIndex,
      blocks: request.blocks,
      attempt_id: request.attemptId,  // attempt 추적용
    });
    if (response.error) throw new Error(response.error.message);
    return {
      blockIndex: response.data!.block_index,
      hintContent: response.data!.hint_content,
      fromCache: response.data!.from_cache,
    };
  },

  /**
   * Get guided hint (단계별 도움, 소크라테스식)
   */
  async getGuidedHint(request: GuidedHintRequest): Promise<GuidedHintResult> {
    const response = await api.post<{
      step_index: number;
      hint_content: string;
      from_cache: boolean;
    }>('/practice/hint/guided', {
      problem_id: request.problemId,
      step_index: request.stepIndex,
      steps: request.steps,
      attempt_id: request.attemptId,  // attempt 추적용
    });
    if (response.error) throw new Error(response.error.message);
    return {
      stepIndex: response.data!.step_index,
      hintContent: response.data!.hint_content,
      fromCache: response.data!.from_cache,
    };
  },

  // ============================================================
  // Guided Tutor Chat (1대1 대화형 튜터)
  // ============================================================

  /**
   * Start guided tutor session (첫 메시지 - 초기 가이드 생성)
   */
  async startGuidedTutor(request: GuidedTutorStartRequest): Promise<GuidedTutorChatResponse> {
    const response = await api.post<{
      response: string;
      is_initial_guide: boolean;
      is_complete: boolean;
      student_progress?: {
        understanding_score: number;
        concepts_mastered: string[];
        concepts_struggling: string[];
        current_focus: string;
        hints_received: number;
        attempts: number;
      };
      initial_guide?: {
        introduction: string;
        concepts_explanation: Array<{
          concept: string;
          explanation: string;
          why_needed: string;
        }>;
        variables_guide: {
          total_count: number;
          variables: Array<{
            role: string;
            type: string;
            initial_value: string;
            suggested_name: string;
            explanation: string;
          }>;
        };
        flow_overview: string[];
        first_step: {
          step_name: string;
          instruction: string;
          starter_question: string;
        };
      };
      suggested_next_step?: string;
    }>('/practice/guided/start', {
      problem_id: request.problemId,
      session_id: request.sessionId,
    }, true, 30000);  // 30초 타임아웃
    if (response.error) throw new Error(response.error.message);
    const data = response.data!;
    return {
      response: data.response,
      isInitialGuide: data.is_initial_guide,
      isComplete: data.is_complete,
      studentProgress: data.student_progress ? {
        understandingScore: data.student_progress.understanding_score,
        conceptsMastered: data.student_progress.concepts_mastered,
        conceptsStruggling: data.student_progress.concepts_struggling,
        currentFocus: data.student_progress.current_focus,
        hintsReceived: data.student_progress.hints_received,
        attempts: data.student_progress.attempts,
      } : undefined,
      initialGuide: data.initial_guide ? {
        introduction: data.initial_guide.introduction,
        conceptsExplanation: data.initial_guide.concepts_explanation.map(c => ({
          concept: c.concept,
          explanation: c.explanation,
          whyNeeded: c.why_needed,
        })),
        variablesGuide: {
          totalCount: data.initial_guide.variables_guide.total_count,
          variables: data.initial_guide.variables_guide.variables.map(v => ({
            role: v.role,
            type: v.type,
            initialValue: v.initial_value,
            suggestedName: v.suggested_name,
            explanation: v.explanation,
          })),
        },
        flowOverview: data.initial_guide.flow_overview,
        firstStep: {
          stepName: data.initial_guide.first_step.step_name,
          instruction: data.initial_guide.first_step.instruction,
          starterQuestion: data.initial_guide.first_step.starter_question,
        },
      } : undefined,
      suggestedNextStep: data.suggested_next_step,
    };
  },

  /**
   * Send chat message to guided tutor (진행 중 대화)
   */
  async chatGuidedTutor(request: GuidedTutorChatRequest): Promise<GuidedTutorChatResponse> {
    const response = await api.post<{
      response: string;
      is_initial_guide: boolean;
      is_complete: boolean;
      student_progress?: {
        understanding_score: number;
        concepts_mastered: string[];
        concepts_struggling: string[];
        current_focus: string;
        hints_received: number;
        attempts: number;
      };
      suggested_next_step?: string;
    }>('/practice/guided/chat', {
      problem_id: request.problemId,
      message: request.message,
      user_code: request.userCode,
      session_id: request.sessionId,
      conversation_history: request.conversationHistory,
    }, true, 30000);  // 30초 타임아웃
    if (response.error) throw new Error(response.error.message);
    const data = response.data!;
    return {
      response: data.response,
      isInitialGuide: data.is_initial_guide,
      isComplete: data.is_complete,
      studentProgress: data.student_progress ? {
        understandingScore: data.student_progress.understanding_score,
        conceptsMastered: data.student_progress.concepts_mastered,
        conceptsStruggling: data.student_progress.concepts_struggling,
        currentFocus: data.student_progress.current_focus,
        hintsReceived: data.student_progress.hints_received,
        attempts: data.student_progress.attempts,
      } : undefined,
      suggestedNextStep: data.suggested_next_step,
    };
  },
};
